src/viam/sdk/components/camera.cpp

```diff
@@ -24,6 +29,106 @@ API API::traits<Camera>::api() {
     return {kRDK, kComponent, "camera"};
 }
 
+namespace {
+// UTF-8 encoding of 'DEPTHMAP' used in the header of FORMAT_RAW_DEPTH bytes payload.
+const uint64_t k_magic_number = 0x44455054484D4150ULL;
+
+// Number of bytes of the header for FORMAT_RAW_DEPTH payloads
+const auto k_header_size = sizeof(uint64_t) * 3;
+
+// Appends a uint64_t value in big-endian format to a byte vector and updates the offset.
+void append_uint64_big_endian(std::vector<unsigned char>& data, size_t* offset, uint64_t value) {
+    const uint64_t value_be = boost::endian::native_to_big(value);
+    if (data.size() < *offset + sizeof(uint64_t)) {
+        throw Exception("Incorrect data size: attempted to write beyond data bounds");
+    }
+    std::memcpy(&data[*offset], &value_be, sizeof(uint64_t));
+    *offset += sizeof(uint64_t);
+}
+
+// Reads a uint64_t value from data in big-endian format and updates the offset.
+// Intended to be used in a sequential manner.
+uint64_t read_uint64_big_endian(const std::vector<unsigned char>& data, size_t* offset) {
+    if (data.size() < *offset + sizeof(uint64_t)) {
+        throw Exception("Attempted to read beyond data bounds.");
+    }
+
+    uint64_t value;
+    std::memcpy(&value, &data[*offset], sizeof(uint64_t));
+    value = boost::endian::big_to_native(value);
+    *offset += sizeof(uint64_t);
+
+    return value;
+}
+
+}  // namespace
+
+std::vector<unsigned char> Camera::encode_depth_map(const Camera::depth_map& m) {
+    const auto& shape = m.shape();
+    if (shape.size() != 2) {
+        throw Exception("Depth map is not a 2D array.");
+    }
+
+    const size_t height = shape[0];
+    const size_t width = shape[1];
+    const size_t total_byte_count = k_header_size + height * width * sizeof(uint16_t);
+    std::vector<unsigned char> data(total_byte_count);
+    size_t offset = 0;
+
+    // Network data is stored in big-endian, while most host systems are little endian.
+    append_uint64_big_endian(data, &offset, k_magic_number);
+    append_uint64_big_endian(data, &offset, width);
+    append_uint64_big_endian(data, &offset, height);
+
+    for (size_t i = 0; i < height; ++i) {
+        for (size_t j = 0; j < width; ++j) {
+            const uint16_t value = m(i, j);
+            const uint16_t value_be = boost::endian::native_to_big(value);
+            std::memcpy(&data[offset], &value_be, sizeof(uint16_t));
+            offset += sizeof(uint16_t);
```

Comment:
If you make `append_uint64_big_endian` a template on the value type, you could re-use it here:

```
template<typename T>
append_big_endian(std::vector<unsigned char>& data, size_t* offset, T value);
```

Reply:
Wow that's so elegant it brings a tear to my eye